crates/testing/src/byzantine/byzantine_behaviour.rs

```diff
@@ -0,0 +1,232 @@
+    /// How many times current node has been elected leader and sent Da Cert
+    pub total_da_certs_sent_from_node: u64,
+    /// Which proposals to be dishonest at
+    pub dishonest_at_da_cert_sent_numbers: HashSet<u64>,
+    /// When leader how many times we will send DacSend and increment view number
+    pub total_views_add_to_cert: u64,
+}
+
+#[async_trait]
+impl<TYPES: NodeType, I: NodeImplementation<TYPES> + std::fmt::Debug, V: Versions>
+    EventTransformerState<TYPES, I, V> for DishonestDa
+{
+    async fn recv_handler(&mut self, event: &HotShotEvent<TYPES>) -> Vec<HotShotEvent<TYPES>> {
+        vec![event.clone()]
+    }
+
+    async fn send_handler(&mut self, event: &HotShotEvent<TYPES>) -> Vec<HotShotEvent<TYPES>> {
+        if let HotShotEvent::DacSend(cert, sender) = event {
+            self.total_da_certs_sent_from_node += 1;
+            if self
+                .dishonest_at_da_cert_sent_numbers
+                .contains(&self.total_da_certs_sent_from_node)
+            {
+                let mut result = vec![HotShotEvent::DacSend(cert.clone(), sender.clone())];
+                for i in 1..=self.total_views_add_to_cert {
+                    let mut bad_cert = cert.clone();
+                    bad_cert.view_number = cert.view_number + i;
+                    result.push(HotShotEvent::DacSend(bad_cert, sender.clone()));
+                }
+                return result;
+            }
+        }
+        vec![event.clone()]
+    }
+}
+
+/// View delay configuration
+#[derive(Debug)]
+pub struct ViewDelay<TYPES: NodeType> {
+    /// How many views the node will be delayed
+    pub number_of_views_to_delay: u64,
+    /// A map that is from view number to vector of events
+    pub events_for_view: HashMap<TYPES::Time, Vec<HotShotEvent<TYPES>>>,
+    /// Specify which view number to stop delaying
+    pub stop_view_delay_at_view_number: u64,
+}
+
+#[async_trait]
+impl<TYPES: NodeType, I: NodeImplementation<TYPES> + std::fmt::Debug, V: Versions>
+    EventTransformerState<TYPES, I, V> for ViewDelay<TYPES>
+{
+    async fn recv_handler(&mut self, event: &HotShotEvent<TYPES>) -> Vec<HotShotEvent<TYPES>> {
+        let correct_event = vec![event.clone()];
+        if let Some(view_number) = event.view_number() {
+            if *view_number >= self.stop_view_delay_at_view_number {
+                return correct_event;
+            }
+
+            // add current view or push event to the map if view number has been added
+            let events_for_current_view = self.events_for_view.entry(view_number).or_default();
+            events_for_current_view.push(event.clone());
+
+            // ensure we are actually able to lookback enough views
+            let view_diff = (*view_number).saturating_sub(self.number_of_views_to_delay);
+            if view_diff > 0 {
+                return match self
+                    .events_for_view
+                    .get(&<TYPES as NodeType>::Time::new(view_diff))
```

Comment:
I feel like this should be a `remove` not a `get`, shouldn't it? Otherwise you'll return all events for a given view multiple times

Reply:
yes good point, we should only need to return these events once

Reply:
if we remove then the next event on the same view will return "correct_event" which we dont want

Reply:
actually, it was flaky returning all the events for every receive, so i added above logic to return remove all events for a given view